jax/_src/lax/lax.py

```diff
@@ -2934,11 +2937,13 @@ def _dot_general_lower(ctx, lhs, rhs, *, dimension_numbers,
                                core.ShapedArray(rhs_aval.shape, aval_out.dtype))
         lhs_dtype = rhs_dtype = aval_out.dtype
     else:  # cpu and gpu
-      lhs = mlir.convert_hlo(ctx, lhs, lhs_aval,
-                             core.ShapedArray(lhs_aval.shape, aval_out.dtype))
-      rhs = mlir.convert_hlo(ctx, rhs, rhs_aval,
-                             core.ShapedArray(rhs_aval.shape, aval_out.dtype))
-      lhs_dtype = rhs_dtype = aval_out.dtype
+      # Do not convert mixed fp8 types to output type.
+      if not _is_fp8_mixed_precision_matmul(lhs_dtype, rhs_dtype):
```

Comment:
At what XLA commit was this feature added? I'm wondering if it was older than the minimum jaxlib version.

Reply:
I'm primarily concerned about backward compatibility. The minimum jaxlib this code supports is 0.4.27 (https://github.com/google/jax/blob/72a81e58e6aa45fb40ca96b9eb7aa9c6c22c3a02/jax/version.py#L136).

Can we confirm this works on both CPU and GPU (the code paths affected) at that jaxlib version?

Reply:
Confirmed after installing
`pip install jaxlib==0.4.27` and ` pip install  https://storage.googleapis.com/jax-releases/cuda12/jaxlib-0.4.27+cuda12.cudnn89-cp310-cp310-manylinux2014_x86_64.whl`
`Jax devices [CpuDevice(id=0)]` and `[cuda(id=0)]` with `jaxlib =0.4.27` return same results as `0.4.29` above. 

Example detailed log:

```
[cuda(id=0)]
0.4.27
{ lambda ; a:f8_e4m3fn[16,64] b:f16[] c:f8_e5m2[64,16] d:f16[]. let
    e:f32[16,16] = pjit[
      name=matmul_fp8
      jaxpr={ lambda ; f:f8_e4m3fn[16,64] g:f16[] h:f8_e5m2[64,16] i:f16[]. let
          j:f16[] = mul g i
          k:bf16[16,16] = dot_general[
            dimension_numbers=(([1], [0]), ([], []))
            preferred_element_type=bfloat16
          ] f h
          l:f32[16,16] = convert_element_type[new_dtype=float32 weak_type=False] k
          m:f32[] = convert_element_type[new_dtype=float32 weak_type=False] j
          n:f32[16,16] = mul l m
        in (n,) }
    ] a b c d
  in (e,) }
module @jit_matmul_fp8 attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<16x64xf8E4M3FN> {mhlo.layout_mode = "default"}, %arg1: tensor<f16> {mhlo.layout_mode = "default"}, %arg2: tensor<64x16xf8E5M2> {mhlo.layout_mode = "default"}, %arg3: tensor<f16> {mhlo.layout_mode = "default"}) -> (tensor<16x16xf32> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.multiply %arg1, %arg3 : tensor<f16>
    %1 = stablehlo.dot_general %arg0, %arg2, contracting_dims = [1] x [0], precision = [DEFAULT, DEFAULT] : (tensor<16x64xf8E4M3FN>, tensor<64x16xf8E5M2>) -> tensor<16x16xbf16>
    %2 = stablehlo.convert %1 : (tensor<16x16xbf16>) -> tensor<16x16xf32>
    %3 = stablehlo.convert %0 : (tensor<f16>) -> tensor<f32>
    %4 = stablehlo.broadcast_in_dim %3, dims = [] : (tensor<f32>) -> tensor<16x16xf32>
    %5 = stablehlo.multiply %2, %4 : tensor<16x16xf32>
    return %5 : tensor<16x16xf32>
  }
}

[[15.400391  15.805664  15.724609  17.183594  16.129883  14.589844
  12.806641  14.589844  17.588867  15.643555  15.5625    14.508789
  14.427734  17.183594  15.886719  14.265625 ]
 [16.535156  16.535156  17.021484  18.723633  17.588867  16.210938
  14.9140625 15.886719  15.724609  17.264648  16.535156  16.454102
  15.076172  17.345703  16.535156  16.697266 ]
 [16.94043   17.021484  15.967773  17.99414   18.480469  15.805664
  13.455078  16.454102  15.724609  17.183594  15.400391  15.481445
  14.103516  17.750977  15.643555  16.61621  ]
 [16.535156  17.10254   15.157227  18.075195  16.535156  17.021484
  13.779297  18.237305  16.859375  17.507812  17.426758  15.319336
  14.9140625 17.507812  16.535156  16.859375 ]
 [16.697266  15.805664  17.10254   17.426758  17.507812  15.967773
  14.265625  17.669922  18.075195  15.886719  15.967773  16.129883
  15.886719  16.77832   15.319336  14.427734 ]
 [17.669922  16.129883  15.805664  17.588867  16.454102  15.481445
  12.563477  15.238281  15.5625    17.183594  14.589844  16.048828
  14.34668   16.291992  14.995117  15.319336 ]
 [17.345703  17.10254   17.669922  18.075195  16.859375  16.373047
  14.103516  16.210938  16.210938  17.913086  17.021484  16.291992
  15.319336  17.10254   16.697266  16.94043  ]
 [17.669922  17.021484  16.77832   17.913086  17.345703  15.724609
  12.887695  16.61621   17.913086  18.480469  15.481445  16.291992
  14.589844  17.426758  16.859375  16.210938 ]
 [16.373047  17.021484  17.021484  17.426758  16.048828  16.535156
  13.374023  16.129883  17.021484  15.967773  16.210938  15.157227
  14.9140625 16.77832   16.697266  15.076172 ]
 [16.373047  14.589844  17.021484  17.021484  16.535156  14.022461
  12.887695  15.076172  14.9140625 13.860352  15.157227  14.103516
  13.6171875 14.022461  13.860352  13.536133 ]
 [14.995117  14.427734  15.238281  17.10254   16.129883  14.995117
  12.3203125 14.427734  14.508789  14.995117  14.508789  13.455078
  13.779297  15.643555  14.508789  14.508789 ]
 [15.805664  15.076172  14.508789  16.77832   15.481445  14.833008
  13.211914  14.670898  14.833008  15.481445  14.103516  14.103516
  13.536133  15.400391  13.779297  13.779297 ]
 [14.589844  13.6171875 14.833008  15.319336  13.455078  13.455078
  12.239258  13.211914  13.374023  13.455078  12.239258  13.698242
  13.130859  13.049805  13.130859  13.860352 ]
 [17.832031  17.913086  17.426758  18.885742  17.183594  17.507812
  14.751953  18.237305  17.426758  19.20996   17.183594  16.77832
  15.643555  16.94043   17.264648  16.535156 ]
 [14.022461  16.048828  15.967773  17.588867  15.481445  15.238281
  14.34668   15.5625    15.643555  15.076172  16.129883  13.049805
  15.319336  15.967773  14.508789  14.589844 ]
 [16.129883  14.751953  15.481445  16.94043   16.048828  14.670898
  13.292969  15.319336  15.5625    16.535156  14.589844  14.833008
  14.995117  16.94043   14.34668   15.076172 ]]
```

```
[CpuDevice(id=0)]
0.4.27
{ lambda ; a:f8_e4m3fn[16,64] b:f16[] c:f8_e5m2[64,16] d:f16[]. let
    e:f32[16,16] = pjit[
      name=matmul_fp8
      jaxpr={ lambda ; f:f8_e4m3fn[16,64] g:f16[] h:f8_e5m2[64,16] i:f16[]. let
          j:f16[] = mul g i
          k:bf16[16,16] = dot_general[
            dimension_numbers=(([1], [0]), ([], []))
            preferred_element_type=bfloat16
          ] f h
          l:f32[16,16] = convert_element_type[new_dtype=float32 weak_type=False] k
          m:f32[] = convert_element_type[new_dtype=float32 weak_type=False] j
          n:f32[16,16] = mul l m
        in (n,) }
    ] a b c d
  in (e,) }
module @jit_matmul_fp8 attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<16x64xf8E4M3FN> {mhlo.layout_mode = "default"}, %arg1: tensor<f16> {mhlo.layout_mode = "default"}, %arg2: tensor<64x16xf8E5M2> {mhlo.layout_mode = "default"}, %arg3: tensor<f16> {mhlo.layout_mode = "default"}) -> (tensor<16x16xf32> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.multiply %arg1, %arg3 : tensor<f16>
    %1 = stablehlo.dot_general %arg0, %arg2, contracting_dims = [1] x [0], precision = [DEFAULT, DEFAULT] : (tensor<16x64xf8E4M3FN>, tensor<64x16xf8E5M2>) -> tensor<16x16xbf16>
    %2 = stablehlo.convert %1 : (tensor<16x16xbf16>) -> tensor<16x16xf32>
    %3 = stablehlo.convert %0 : (tensor<f16>) -> tensor<f32>
    %4 = stablehlo.broadcast_in_dim %3, dims = [] : (tensor<f32>) -> tensor<16x16xf32>
    %5 = stablehlo.multiply %2, %4 : tensor<16x16xf32>
    return %5 : tensor<16x16xf32>
  }
}

[[15.383043  15.798857  15.691443  17.154228  16.10875   14.577179
  12.778343  14.585698  17.573433  15.652618  15.583159  14.499221
  14.398588  17.201244  15.892665  14.260005 ]
 [16.568085  16.538086  17.018793  18.710691  17.556175  16.200014
  14.937374  15.916076  15.690019  17.284912  16.56096   16.49176
  15.084666  17.346312  16.499142  16.693386 ]
 [16.962791  17.022118  15.938882  18.001265  18.509756  15.829094
  13.421358  16.47485   15.744477  17.164043  15.398728  15.447787
  14.121206  17.7694    15.645375  16.597214 ]
 [16.53864   17.079742  15.162767  18.076818  16.499537  17.004545
  13.793386  18.25174   16.899033  17.53235   17.458063  15.303495
  14.887518  17.49403   16.5666    16.831038 ]
 [16.73428   15.804714  17.140059  17.424145  17.516598  15.9559
  14.26606   17.707674  18.082933  15.899185  15.956884  16.163305
  15.887436  16.820114  15.343201  14.406204 ]
 [17.699526  16.130674  15.787617  17.61582   16.428455  15.477171
  12.577487  15.249076  15.560759  17.189728  14.625938  16.02149
  14.384682  16.258905  14.9586115 15.324402 ]
 [17.30565   17.077843  17.71045   18.09166   16.837212  16.401443
  14.08436   16.176165  16.248615  17.949734  16.987864  16.281336
  15.290296  17.137823  16.65842   16.913517 ]
 [17.71037   17.031616  16.774124  17.90604   17.348553  15.7103615
  12.925531  16.64051   17.897335  18.471762  15.5157585 16.258312
  14.616569  17.415636  16.849085  16.227085 ]
 [16.355177  17.007395  16.999361  17.449396  16.037073  16.512123
  13.35748   16.104538  17.045944  15.996232  16.25085   15.14617
  14.93957   16.801353  16.698135  15.070384 ]
 [16.370989  14.595246  16.989513  16.998014  16.521868  14.035205
  12.849929  15.060979  14.877216  13.886988  15.14753   14.086409
  13.591234  13.99258   13.874876  13.576531 ]
 [15.023588  14.420235  15.268217  17.075153  16.162159  15.002103
  12.320184  14.4074955 14.471437  14.95779   14.472032  13.458294
  13.797898  15.636352  14.483559  14.520574 ]
 [15.838988  15.059668  14.502477  16.747133  15.509862  14.868291
  13.237204  14.677998  14.8354225 15.4631605 14.071062  14.1116295
  13.518813  15.364583  13.804369  13.754363 ]
 [14.605096  13.651224  14.85438   15.294986  13.4266615 13.482506
  12.260195  13.176955  13.394525  13.491924  12.264993  13.6735115
  13.1418    13.061381  13.162363  13.872898 ]
 [17.83591   17.93145   17.420822  18.899199  17.152882  17.541294
  14.751795  18.20949   17.402536  19.212177  17.214622  16.813702
  15.623202  16.91425   17.305809  16.521383 ]
 [14.061306  16.047918  15.965675  17.561455  15.481762  15.26652
  14.380395  15.53278   15.619313  15.103817  16.163519  13.067931
  15.304667  15.948935  14.479542  14.6296   ]
 [16.129963  14.787256  15.461419  16.91977   16.062363  14.654546
  13.270132  15.360768  15.561432  16.50658   14.550373  14.822164
  15.031533  16.955984  14.316957  15.097148 ]]
```